src/reducers/components.ts

```diff
@@ -0,0 +1,142 @@
+import { INITIALIZATION, SET_CURRENT_COMPONENT, UPDATE_GRAPHIC, CLEAR_FINAL_STATES, SET_AUTO_CLEAR, GlobalComponentsAction } from "actions/components";
+import * as go from "gojs";
+import { DrawComponent } from "utils/drawComponent";
+import { modelTags } from "utils/configurationParser";
+import { activeStateColor, stateColor } from "utils/graphicColors";
+import { Reducer } from "redux";
+
+export interface Instance {
+    jsonMessage: any;
+    stateMachineRef: any;
+    isFinal: boolean;
+};
+
+export interface ComponentProperties {
+    diagram: go.Diagram;
+    stateMachineProperties: { [name: string]: { [id: number]: Instance } };
+    finalStates: Array<String>;
+    entryPointState: string;
+};
+
+export interface ComponentsState {
+    componentProperties: { [componentName: string]: ComponentProperties };
+    currentComponent: string;
+    projectName: string;
+    initialized: boolean;
+    autoClear: boolean;
+};
+
+const initialState: ComponentsState = {
+    componentProperties: {},
+    currentComponent: undefined,
+    projectName: undefined,
+    initialized: false,
+    autoClear: false
+};
+
+const updateState = (diagram, stateKey, finalStates, entryPointState, increment) => {
+    let data = diagram.findNodeForKey(stateKey).data;
+    let oldValue = data.numberOfStates;
+    let newValue = oldValue + increment;
+    diagram.model.setDataProperty(data, "numberOfStates", newValue);
+    diagram.model.setDataProperty(data, "text", data.stateName + " (" + data.numberOfStates + ")");
+    if (finalStates.indexOf(stateKey) > -1 || entryPointState === stateKey)
+        return;
+    if (newValue === 0) {
+        diagram.model.setDataProperty(data, "fill", stateColor);
+        diagram.model.setDataProperty(data, "stroke", stateColor);
+    } else {
+        diagram.model.setDataProperty(data, "fill", activeStateColor);
+        diagram.model.setDataProperty(data, "stroke", activeStateColor);
+    }
+};
+
+const clearFinalStates = (diagram, finalStatesToClear, stateMachine, numberOfInstances) => {
+    diagram.model.startTransaction(CLEAR_FINAL_STATES);
+    let stateMachineData = diagram.findNodeForKey(stateMachine).data;
+    diagram.model.setDataProperty(stateMachineData, "numberOfInstances", numberOfInstances);
+    diagram.model.setDataProperty(stateMachineData, "text", stateMachineData.key + " (" + stateMachineData.numberOfInstances + ")");
```

Comment:
optional: you can use string interpolation such as `(${stateMachineData.numberOfInstance})`